<a href="https://colab.research.google.com/github/imxenia/PiJ/blob/master/%D1%88%D0%B0%D0%B3_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Лышова К.А.

In [1]:
!pip install --upgrade gdown
!gdown --id 1SQihqE13h5sEVzLSaBn9f2l_FdMtxNQb

/usr/local/lib/python3.10/dist-packages/gdown/__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=1SQihqE13h5sEVzLSaBn9f2l_FdMtxNQb
To: /content/movies.csv
100% 1.89M/1.89M [00:00<00:00, 131MB/s]


In [2]:
import pandas as pd
dataset = pd.read_csv('movies.csv')
dataset.head()

poster  \
0  https://m.media-amazon.com/images/S/sash/4Fyxw...   
1  https://m.media-amazon.com/images/S/sash/4Fyxw...   
2  https://m.media-amazon.com/images/S/sash/4Fyxw...   
3  https://m.media-amazon.com/images/S/sash/4Fyxw...   
4  https://m.media-amazon.com/images/S/sash/4Fyxw...   

                      title certificate  runtime                      genre  \
0  The Shawshank Redemption          15  142 min                      Drama   
1           The Dark Knight         12A  152 min       Action, Crime, Drama   
2                 Inception         12A  148 min  Action, Adventure, Sci-Fi   
3                Fight Club          18  139 min                      Drama   
4              Pulp Fiction          18  154 min               Crime, Drama   

   rating                                              about  \
0     9.3  Two imprisoned men bond over a number of years...   
1     9.0  When the menace known as the Joker wreaks havo...   
2     8.8  A thief who steals corporate secrets through t...   
3     8.8  An insomniac office worker and a devil-may-car...   
4     8.9  The lives of two mob hitmen, a boxer, a gangst...   

            director                                              stars  \
0     Frank Darabont  ('Tim Robbins',), ('Morgan Freeman',), ('Bob G...   
1  Christopher Nolan  ('Christian Bale',), ('Heath Ledger',), ('Aaro...   
2  Christopher Nolan  ('Leonardo DiCaprio',), ('Joseph Gordon-Levitt...   
3      David Fincher  ('Brad Pitt',), ('Edward Norton',), ('Meat Loa...   
4  Quentin Tarantino  ('John Travolta',), ('Uma Thurman',), ('Samuel...   

       votes gross_earn  
0  26,26,905    $28.34M  
1  25,98,173   $534.86M  
2  23,04,062   $292.58M  
3  20,71,088    $37.03M  
4  20,11,013   $107.93M

In [3]:
dataset.shape

(5000, 11)

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   poster       5000 non-null   object 
 1   title        5000 non-null   object 
 2   certificate  5000 non-null   object 
 3   runtime      5000 non-null   object 
 4   genre        4970 non-null   object 
 5   rating       5000 non-null   float64
 6   about        4986 non-null   object 
 7   director     5000 non-null   object 
 8   stars        5000 non-null   object 
 9   votes        5000 non-null   object 
 10  gross_earn   4585 non-null   object 
dtypes: float64(1), object(10)
memory usage: 429.8+ KB


In [5]:
dataset.duplicated().sum()

0

In [6]:
dataset = dataset.drop_duplicates()

In [7]:
dataset['gross_earn'].isnull().sum()

415

In [8]:
dataset['runtime']

0              142 min
1              152 min
2              148 min
3              139 min
4              154 min
             ...      
4995           102 min
4996            94 min
4997    Drama, Romance
4998            92 min
4999           120 min
Name: runtime, Length: 5000, dtype: object

In [9]:
dataset['runtime'].unique()

array(['142 min', '152 min', '148 min', '139 min', '154 min', '178 min',
       '136 min', '201 min', '175 min', '164 min', '179 min', '127 min',
       '143 min', '155 min', '140 min', '165 min', '169 min', '121 min',
       '118 min', '162 min', '153 min', '151 min', '195 min', '113 min',
       '130 min', '124 min', '122 min', '194 min', '132 min', '202 min',
       '138 min', '119 min', '189 min', '180 min', '137 min', '106 min',
       '111 min', '116 min', '145 min', '110 min', '98 min', '107 min',
       '126 min', '100 min', '131 min', '133 min', '103 min', '108 min',
       '96 min', '99 min', '115 min', '88 min', '120 min', '81 min',
       '135 min', '146 min', '117 min', '104 min', '92 min', '91 min',
       '102 min', '149 min', '170 min', '112 min', '101 min', '141 min',
       '144 min', '114 min', '150 min', '90 min', '161 min', '128 min',
       '93 min', '129 min', '166 min', '147 min', '134 min', '109 min',
       '105 min', '125 min', '156 min', '123 min', '95 min',

In [10]:
import re

In [25]:
import numpy as np

In [11]:
def clear_runtime(runtime):
    return re.sub(r'\D', '', runtime)

In [12]:
dataset['runtime_clear'] = dataset['runtime'].apply(clear_runtime)

In [13]:
dataset['runtime_clear']

0       142
1       152
2       148
3       139
4       154
       ... 
4995    102
4996     94
4997       
4998     92
4999    120
Name: runtime_clear, Length: 5000, dtype: object

In [14]:
dataset['runtime_clear'].isna().sum()

0

In [15]:
dataset['runtime_clear'][4997]

''

In [16]:
type(dataset['runtime_clear'][4997])

str

In [17]:
dataset = dataset[dataset['runtime_clear'] != '']

In [18]:
dataset.shape

(4970, 12)

In [19]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4970 entries, 0 to 4999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   poster         4970 non-null   object 
 1   title          4970 non-null   object 
 2   certificate    4970 non-null   object 
 3   runtime        4970 non-null   object 
 4   genre          4970 non-null   object 
 5   rating         4970 non-null   float64
 6   about          4956 non-null   object 
 7   director       4970 non-null   object 
 8   stars          4970 non-null   object 
 9   votes          4970 non-null   object 
 10  gross_earn     4574 non-null   object 
 11  runtime_clear  4970 non-null   object 
dtypes: float64(1), object(11)
memory usage: 504.8+ KB


In [20]:
dataset['runtime_clear'] = dataset['runtime_clear'].astype(int)

In [21]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4970 entries, 0 to 4999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   poster         4970 non-null   object 
 1   title          4970 non-null   object 
 2   certificate    4970 non-null   object 
 3   runtime        4970 non-null   object 
 4   genre          4970 non-null   object 
 5   rating         4970 non-null   float64
 6   about          4956 non-null   object 
 7   director       4970 non-null   object 
 8   stars          4970 non-null   object 
 9   votes          4970 non-null   object 
 10  gross_earn     4574 non-null   object 
 11  runtime_clear  4970 non-null   int64  
dtypes: float64(1), int64(1), object(10)
memory usage: 504.8+ KB


In [22]:
dataset['gross_earn']

0        $28.34M
1       $534.86M
2       $292.58M
3        $37.03M
4       $107.93M
          ...   
4994         NaN
4995      $9.40M
4996         NaN
4998      $6.67M
4999      $8.20M
Name: gross_earn, Length: 4970, dtype: object

In [23]:
def clear_gross_earn(gross_earn):
    if pd.isna(gross_earn):
        return np.nan
    else:
        return re.sub('[^\d.]', '', gross_earn)

In [26]:
dataset['gross_earn_clear'] = dataset['gross_earn'].apply(clear_gross_earn)

In [27]:
dataset['gross_earn_clear']

0        28.34
1       534.86
2       292.58
3        37.03
4       107.93
         ...  
4994       NaN
4995      9.40
4996       NaN
4998      6.67
4999      8.20
Name: gross_earn_clear, Length: 4970, dtype: object

In [28]:
dataset['gross_earn_clear'] = dataset['gross_earn_clear'].astype(float)

In [29]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4970 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   poster            4970 non-null   object 
 1   title             4970 non-null   object 
 2   certificate       4970 non-null   object 
 3   runtime           4970 non-null   object 
 4   genre             4970 non-null   object 
 5   rating            4970 non-null   float64
 6   about             4956 non-null   object 
 7   director          4970 non-null   object 
 8   stars             4970 non-null   object 
 9   votes             4970 non-null   object 
 10  gross_earn        4574 non-null   object 
 11  runtime_clear     4970 non-null   int64  
 12  gross_earn_clear  4574 non-null   float64
dtypes: float64(2), int64(1), object(10)
memory usage: 543.6+ KB


In [30]:
dataset['gross_earn_clear']

0        28.34
1       534.86
2       292.58
3        37.03
4       107.93
         ...  
4994       NaN
4995      9.40
4996       NaN
4998      6.67
4999      8.20
Name: gross_earn_clear, Length: 4970, dtype: float64

In [31]:
sorted_dataset = dataset.sort_values(by='rating', ascending=False)

In [32]:
sorted_dataset.head()

poster  \
0     https://m.media-amazon.com/images/S/sash/4Fyxw...   
3788  https://m.media-amazon.com/images/S/sash/4Fyxw...   
9     https://m.media-amazon.com/images/S/sash/4Fyxw...   
32    https://m.media-amazon.com/images/S/sash/4Fyxw...   
24    https://m.media-amazon.com/images/S/sash/4Fyxw...   

                         title certificate  runtime  \
0     The Shawshank Redemption          15  142 min   
3788            Hababam Sinifi         12A   87 min   
9                The Godfather           X  175 min   
32      The Godfather: Part II           X  202 min   
24            Schindler's List          15  195 min   

                          genre  rating  \
0                         Drama     9.3   
3788              Comedy, Drama     9.2   
9                  Crime, Drama     9.2   
32                 Crime, Drama     9.0   
24    Biography, Drama, History     9.0   

                                                  about              director  \
0     Two imprisoned men bond over a number of years...        Frank Darabont   
3788  Lazy, uneducated students share a very close b...         Ertem Egilmez   
9     The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
32    The early life and career of Vito Corleone in ...  Francis Ford Coppola   
24    In German-occupied Poland during World War II,...      Steven Spielberg   

                                                  stars      votes gross_earn  \
0     ('Tim Robbins',), ('Morgan Freeman',), ('Bob G...  26,26,905    $28.34M   
3788  ('Kemal Sunal',), ('Münir Özkul',), ('Halit Ak...     40,492        NaN   
9     ('Marlon Brando',), ('Al Pacino',), ('James Ca...  18,19,641   $134.97M   
32    ('Al Pacino',), ('Robert De Niro',), ('Robert ...  12,49,182    $57.30M   
24    ('Liam Neeson',), ('Ralph Fiennes',), ('Ben Ki...  13,33,407    $96.90M   

      runtime_clear  gross_earn_clear  
0               142             28.34  
3788             87               NaN  
9               175            134.97  
32              202             57.30  
24              195             96.90

In [33]:
sorted_dataset.reset_index(drop=True, inplace=True)

In [34]:
sorted_dataset.head()

poster  \
0  https://m.media-amazon.com/images/S/sash/4Fyxw...   
1  https://m.media-amazon.com/images/S/sash/4Fyxw...   
2  https://m.media-amazon.com/images/S/sash/4Fyxw...   
3  https://m.media-amazon.com/images/S/sash/4Fyxw...   
4  https://m.media-amazon.com/images/S/sash/4Fyxw...   

                      title certificate  runtime                      genre  \
0  The Shawshank Redemption          15  142 min                      Drama   
1            Hababam Sinifi         12A   87 min              Comedy, Drama   
2             The Godfather           X  175 min               Crime, Drama   
3    The Godfather: Part II           X  202 min               Crime, Drama   
4          Schindler's List          15  195 min  Biography, Drama, History   

   rating                                              about  \
0     9.3  Two imprisoned men bond over a number of years...   
1     9.2  Lazy, uneducated students share a very close b...   
2     9.2  The aging patriarch of an organized crime dyna...   
3     9.0  The early life and career of Vito Corleone in ...   
4     9.0  In German-occupied Poland during World War II,...   

               director                                              stars  \
0        Frank Darabont  ('Tim Robbins',), ('Morgan Freeman',), ('Bob G...   
1         Ertem Egilmez  ('Kemal Sunal',), ('Münir Özkul',), ('Halit Ak...   
2  Francis Ford Coppola  ('Marlon Brando',), ('Al Pacino',), ('James Ca...   
3  Francis Ford Coppola  ('Al Pacino',), ('Robert De Niro',), ('Robert ...   
4      Steven Spielberg  ('Liam Neeson',), ('Ralph Fiennes',), ('Ben Ki...   

       votes gross_earn  runtime_clear  gross_earn_clear  
0  26,26,905    $28.34M            142             28.34  
1     40,492        NaN             87               NaN  
2  18,19,641   $134.97M            175            134.97  
3  12,49,182    $57.30M            202             57.30  
4  13,33,407    $96.90M            195             96.90

In [44]:
sorted_dataset.to_csv('movies_clean.csv', index=False)

In [46]:
import os

# Получаем текущую рабочую директорию
current_directory = os.getcwd()

# Полный путь к файлу
file_path = os.path.join(current_directory, 'movies_clean.csv')

print("Полный путь к файлу:", file_path)

Полный путь к файлу: /content/movies_clean.csv


Вывод:

С помощью dataset.shape можно понять, сколько строк и столбцов в датасете. В данном случае количество  наблюдений - 5000, количество признаков - 11.

В датасете присутствуют следующие типы данных:
object (строки) и float64.

Предобработка типов данных:
Признаки runtime, votes, и gross_earn представлены как строки (object), хотя это числовые значения. Эти столбцы нужно привести к числовому типу, чтобы с ними можно было проводить арифметические операции.
Также нужно убедиться, что нет лишних символов в этих столбцах, таких как "часов" или "минут" в runtime, которые нужно будет удалить для корректной работы с данными.

Дубликаты:
В датасете нет дубликатов, что подтверждается результатом dataset.duplicated().sum(), равным 0.

Предобработка:
Так как дубликатов нет, дополнительная обработка на этом этапе не требуется.

Пропущенные значения:
В столбце genre отсутствуют данные для 30 фильмов (5000 - 4970 = 30).
В столбце about отсутствуют данные для 14 фильмов (5000 - 4986 = 14).
В столбце gross_earn отсутствуют данные для 415 фильмов (5000 - 4585 = 415).

Предобработка пропусков:
Пропуски в столбце gross_earn составляют значительный процент (415 из 5000), что может существенно повлиять на анализ. Для их обработки можно использовать медиану или среднее значение, либо оставить пропуски, если они не критичны для конкретного анализа.
Пропуски в столбцах genre и about можно заполнить наиболее часто встречающимся значением или оставить, если это не оказывает значимого влияния на результаты.

О возможных проблемах, которые могут возникнуть при анализе:
Преобразование столбцов runtime, votes и gross_earn в числовые значения потребует предварительной очистки данных. Например, в столбце gross_earn могут быть строковые символы в значениях, которые нужно удалить.
Пропуски в столбце gross_earn могут быть связаны с отсутствием информации о доходах некоторых фильмов, что может повлиять на анализ по доходности.
Возможно, в столбцах runtime и votes будут присутствовать значения, которые не соответствуют ожидаемым формам. Эти проблемы могут потребовать дополнительной очистки.
В столбце genre могут быть указаны несколько жанров для одного фильма, что потребует дополнительной обработки для правильной работы с этим признаком, например, разделение жанров по запятой.